Add automated Git configuration, Nerd Fonts installer, and bootloader management scripts to linutil

fam007e opened the pull request:
## Type of Change
- [x] New feature
- [ ] Bug fix
- [ ] Documentation Update
- [ ] Refactoring
- [ ] Hotfix
- [ ] Security patch
- [x] UI/UX improvement

## Description
This pull request introduces several new features to **linutil**:

1. **Automated Git Configuration Script**: This script helps users set up Git by configuring their GitHub email, SSH key (either Ed25519 or RSA), and copying the SSH key to the clipboard for easy GitHub integration. It checks for necessary tools like `xclip` and uses best practices for SSH key generation and GitHub authentication.

2. **Nerd Fonts Installer Script**: This script automates the installation of Nerd Fonts by detecting the user’s Linux distribution and using the appropriate package manager to install dependencies (`curl`, `tar`). It then provides an interactive prompt for selecting fonts, downloads them from GitHub, installs them locally, and refreshes the font cache. This feature simplifies the process of installing Nerd Fonts across different systems.

3. **Bootloader Switching and Backup Script**: This new script facilitates seamless switching between GRUB and systemd-boot bootloaders. It includes:
   - **Interactive Menu**: Detects the current bootloader and offers options to switch to the other bootloader, restore from backups, or exit.
   - **Backup and Restore**: Backs up GRUB or systemd-boot configurations and allows restoring from previous backups.
   - **Automatic Detection**: Automatically detects the current bootloader and provides appropriate actions based on the detected bootloader.

4. **Edited Logo**: Used ANSI color from `ratatui` style's color and modifier to create a logo which is in depicted in readme.

5. Minor correction of word "parent" in `hint.rs` is made. 

These changes aim to streamline user setup and enhance the configuration experience, making it easier for developers to quickly configure their environment and manage bootloader setups.

## Testing
- The Git configuration script has been tested on various distributions (Ubuntu, Fedora, Arch) for SSH key generation, GitHub key addition, and correct handling of user input for key type and passphrase.
- The Nerd Fonts installer script has been tested across multiple distributions to ensure it correctly installs fonts, refreshes the font cache, and handles package dependencies.
- The Bootloader Switching and Backup script has been tested for interactive menu functionality, accurate detection of the current bootloader, and successful switching between `GRUB` and `systemd-boot`. Backup and restore functionalities have been verified.

## Impact
- **Performance**: Minimal performance impact since all scripts run only during setup or maintenance processes.
- **Dependencies**: Introduces requirements for `xclip` in the Git configuration script, and ensures `curl` and `tar` are installed for the Nerd Fonts installer. The bootloader script uses common system tools and does not introduce additional runtime dependencies. 
- **Behavior**: Enhances the user setup experience by automating Git configuration, font installation, and providing a streamlined process for managing bootloaders.
- **Colorful**: Logo with ASCII art is colored adds to its visual appeal.  

## Issue related to PR
- Resolves #256
- Resolves #307 
- Resolves #160 
- Resolves #308

## Additional Information
- The Git configuration script includes detailed user prompts and instructions for adding SSH keys to GitHub.
- The Nerd Fonts installer script uses modular, reusable functions for OS detection and dependency installation, ensuring it is extensible for future font management tasks.
- The Bootloader Switching and Backup script includes comprehensive backup and restoration processes, as well as an intuitive interactive menu for switching and restoring bootloaders.

## Checklist
- [x] My code adheres to the coding and style guidelines of the project.
- [x] I have performed a self-review of my own code.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the documentation.
- [x] My changes generate no errors/warnings/merge conflicts.

nnyyxxxx commented:
Why are you removing cargo.lock?

nnyyxxxx commented on src/commands/system-setup/5-git-config-auto.sh:
```diff
@@ -0,0 +1,78 @@
+
+    # If user opts for a passphrase, add key to SSH agent
+    if [ "$use_passphrase" = "y" ]; then
+        ssh-add -l &>/dev/null || eval "$(ssh-agent -s)"
```
```bash
ssh-add -l &>/dev/null || eval "$(ssh-agent -s)"
```
This is not posix compliant.

fam007e replied:
## Key Changes:

1. Removed read `-p`: POSIX `sh` doesn’t support the `-p` option for `read`, so it has been replaced with `printf` and plain read.
2. Removed Bash-specific redirection (`&>/dev/null`): Replaced with POSIX-compliant `>/dev/null 2>&1`.
3. Command existence checks: `command -v xclip` is used to check for `xclip` existence instead of relying on assumptions.
4. Shell-specific behaviors: Avoided non-POSIX constructs like function keywords without parentheses (`function setup_git_config`).

nnyyxxxx commented on src/commands/system-setup/6-nerd-fonts-setup.sh:
```diff
@@ -0,0 +1,122 @@
+    echo "Select fonts to install (separate with spaces):"
+    echo "---------------------------------------------"
+    for i in "${!fonts[@]}"; do
+        echo " $i  -  ${fonts[i]}"
```
```bash
echo " $i  -  ${fonts[i]}"
```
This is not posix compliant.

nnyyxxxx commented on src/commands/system-setup/6-nerd-fonts-setup.sh:
```diff
@@ -0,0 +1,122 @@
+# Function to download and install the selected fonts
+download_and_install_fonts() {
+    for selection in $font_selection; do
+        font=${fonts[$selection]}
```
```bash
font=${fonts[$selection]}
```
This is not posix compliant.

fam007e replied:
### Key Changes:

- **Removed Bash arrays**: Replaced the array syntax (`fonts=()`) with a space-separated string and looped over it with `for`.
- **Replaced `read -rp`**: Replaced with a combination of `printf` and `read` for POSIX compliance.
- **Handling font selections**: Used basic `sh` syntax to loop through the string of fonts and matched them with user input.

nnyyxxxx commented on src/commands/system-setup/5-git-config-auto.sh:
```diff
@@ -0,0 +1,88 @@
+#!/bin/sh
```
```suggestion
#!/bin/sh -e
```

fam007e replied:
I have already done it in recent commit.

nnyyxxxx replied:
@fam007e You have not.

nnyyxxxx commented on src/commands/system-setup/6-nerd-fonts-setup.sh:
```diff
@@ -0,0 +1,131 @@
+#!/bin/sh
```
```suggestion
#!/bin/sh -e
```

fam007e commented:
I hope the commits have fixed the issue.
BTW I saw a recent cargo build warning due to a new `ratatui` version...

`cargo run` warning mesage shown:
```sh
warning: use of deprecated method `ratatui::Frame::<'_>::set_cursor`: the method set_cursor_position indicates more clearly what about the cursor to set
   --> src/filter.rs:111:19
    |
111 |             frame.set_cursor(x, y);
    |                   ^^^^^^^^^^
    |
    = note: `#[warn(deprecated)]` on by default

warning: use of deprecated method `ratatui::Frame::<'_>::size`: use .area() as it's the more correct name
  --> src/state.rs:81:26
   |
81 |             .split(frame.size());
   |                          ^^^^

warning: `tui` (bin "linutil") generated 2 warnings
```

nnyyxxxx commented on src/commands/system-setup/6-nerd-fonts-setup.sh:
```diff
@@ -0,0 +1,131 @@
+check_environment() {
+    echo "Checking environment..."
+    checkEnv  # Assuming this is a function from common-script.sh
+}
```
```bash
# Function to check environment and dependencies
check_environment() {
    echo "Checking environment..."
    checkEnv  # Assuming this is a function from common-script.sh
}
```
This needs to be removed.

nnyyxxxx commented on src/commands/system-setup/6-nerd-fonts-setup.sh:
```diff
@@ -0,0 +1,131 @@
+}
+
+# Main script execution
+check_environment
```
```bash
check_environment
```

This needs to be renamed to checkEnv

nnyyxxxx commented on src/commands/system-setup/5-git-config-auto.sh:
```diff
@@ -0,0 +1,88 @@
+}
+
+# Check environment and necessary tools
+checkEnv
```
```bash
checkEnv
```
This needs to be moved under ``# Main execution``

nnyyxxxx replied:
@fam007e checkEnv still needs to be moved under ``# Main execution``

nnyyxxxx commented:
@fam007e These need to be put on separate lines with ``Resolves`` before them.
![image](https://github.com/user-attachments/assets/ccce2cab-4fce-44a8-a9d3-93059a802f77)

nnyyxxxx commented on src/commands/system-setup/7-bootloader-switch.sh:
```diff
@@ -0,0 +1,170 @@
+checkEnv
+
+# Run the menu function for user interaction
+menu
```
```bash
# Check environment and prerequisites
checkEnv

# Run the menu function for user interaction
menu
```
These need to be combined and put under eachother at the end of the script and you have to call ``checkEscalationTool`` aswell.

fam007e replied:
I have fixed these by my recent commits. Thanks for pointing them out.

nnyyxxxx commented on src/commands/system-setup/6-nerd-fonts-setup.sh:
```diff
@@ -0,0 +1,125 @@
+                checkCommandRequirements "tar"
+
+                # Download the font
+                wget -q --show-progress "https://github.com/ryanoasis/nerd-fonts/releases/latest/download/$font_name.tar.xz" -P "$HOME/tmp"
```
Use curl -sSLo instead of wget please.

fam007e replied:
Fixed. May I know why `wget` is not favored in this instance? Was that to do with `posix`compliance?

nnyyxxxx commented:
@fam007e You're adding to many things into 1 PR I'd recommend splitting this up into multiple.

fam007e commented:
OK I won't add anymore to this.

nnyyxxxx commented on src/commands/system-setup/arch/cosmic-setup.sh:
```diff
@@ -0,0 +1,119 @@
+#!/bin/sh -e
+
+# Source the common script
+. "../../common-script.sh"
```
No need for these quotation marks here.

nnyyxxxx commented on src/commands/system-setup/arch/cosmic-setup.sh:
```diff
@@ -0,0 +1,119 @@
+checkEnv
+checkEscalationTool
+check_gpu_cosmic_compatibility
+install_aur_helper
```
```bash
# Main execution
checkEnv
checkEscalationTool
check_gpu_cosmic_compatibility
install_aur_helper
```
This needs to be at the bottom of the script.

nnyyxxxx commented on src/commands/system-setup/arch/cosmic-setup.sh:
```diff
@@ -0,0 +1,119 @@
+# Toolkit integration
+install_aur_helper() {
+    . "./paru-setup.sh" || . "./yay-setup.sh"
+}
```
```bash
# Toolkit integration
install_aur_helper() {
    . "./paru-setup.sh" || . "./yay-setup.sh"
}
``` 
No need for this.

nnyyxxxx commented on src/commands/system-setup/arch/cosmic-setup.sh:
```diff
@@ -0,0 +1,119 @@
+
+    $ESCALATION_TOOL systemctl enable cosmic-greeter.service
+    xdg-user-dirs-update
+    $ESCALATION_TOOL systemctl enable power-profiles-daemon.service
```
```bash
    $ESCALATION_TOOL systemctl enable power-profiles-daemon.service
```
This is being called twice.

nnyyxxxx commented on src/commands/system-setup/arch/cosmic-setup.sh:
```diff
@@ -0,0 +1,119 @@
+
+echo "Installing Bluetooth packages..."
+$ESCALATION_TOOL pacman -S --needed --noconfirm bluez bluez-utils bluez-plugins bluez-hid2hci bluez-cups bluez-libs bluez-tools
+$ESCALATION_TOOL systemctl enable bluetooth.service
```
```bash

echo "Installing PipeWire packages..."
$ESCALATION_TOOL pacman -S --needed --noconfirm gstreamer gst-libav gst-plugins-bad gst-plugins-base gst-plugins-good gst-plugins-ugly libdvdcss alsa-utils alsa-firmware pavucontrol lib32-pipewire-jack libpipewire pipewire-v4l2 pipewire-x11-bell pipewire-zeroconf realtime-privileges sof-firmware ffmpeg ffmpegthumbs ffnvcodec-headers

echo "Installing Bluetooth packages..."
$ESCALATION_TOOL pacman -S --needed --noconfirm bluez bluez-utils bluez-plugins bluez-hid2hci bluez-cups bluez-libs bluez-tools
$ESCALATION_TOOL systemctl enable bluetooth.service
```
All of this is redundant, this script should only be installing cosmic and dependencies; nothing else.

nnyyxxxx commented on src/commands/system-setup/arch/cosmic-setup.sh:
```diff
@@ -0,0 +1,119 @@
+    $ESCALATION_TOOL grub-mkconfig -o /boot/grub/grub.cfg
+else
+    echo "GRUB not found. Skipping OS-Prober."
+fi
```
```bash
# Check if GRUB is installed and add OS-Prober support
if command -v grub-mkconfig >/dev/null 2>&1; then
    echo "GRUB is installed. Adding support for OS-Prober."
    $ESCALATION_TOOL pacman -S --needed --noconfirm os-prober
    $ESCALATION_TOOL sed -i 's/#[[:space:]]*GRUB_DISABLE_OS_PROBER=false/GRUB_DISABLE_OS_PROBER=false/' /etc/default/grub
    $ESCALATION_TOOL os-prober
    $ESCALATION_TOOL grub-mkconfig -o /boot/grub/grub.cfg
else
    echo "GRUB not found. Skipping OS-Prober."
fi
```
All of this is redundant.

nnyyxxxx commented on src/commands/system-setup/arch/cosmic-setup.sh:
```diff
@@ -0,0 +1,119 @@
+    *)
+        echo "Not running in a VM."
+        ;;
+esac
```
```bash
# VM detection and relevant package installation
VM_TYPE=$(systemd-detect-virt)
case "$VM_TYPE" in
    oracle)
        $ESCALATION_TOOL pacman -S --needed --noconfirm virtualbox-guest-utils
        ;;
    kvm)
        $ESCALATION_TOOL pacman -S --needed --noconfirm qemu-guest-agent spice-vdagent
        ;;
    vmware)
        $ESCALATION_TOOL pacman -S --needed --noconfirm xf86-video-vmware open-vm-tools xf86-input-vmmouse
        $ESCALATION_TOOL systemctl enable vmtoolsd.service
        ;;
    *)
        echo "Not running in a VM."
        ;;
esac
```
All of this is redundant.

nnyyxxxx commented on src/commands/system-setup/arch/cosmic-setup.sh:
```diff
@@ -0,0 +1,119 @@
+    else
+        echo "Cannot detect GPU. 'lspci' command not found."
+    fi
+}
```
```bash
# Check GPU compatibility for Wayland
check_gpu_cosmic_compatibility() {
    if command -v lspci >/dev/null 2>&1; then
        GPU_INFO=$(lspci | grep -E "VGA|3D")

        case "$GPU_INFO" in
            *NVIDIA*)
                if echo "$GPU_INFO" | grep -Eq "GTX (9[0-9]{2}|[1-9][0-9]{3})|RTX|Titan|A[0-9]{2,3}"; then
                    echo "Your NVIDIA GPU should support Cosmic Desktop."
                else
                    echo "Older NVIDIA GPU detected. Only 900 series and newer support Cosmic Desktop."
                fi
                ;;
            *Intel*)
                if echo "$GPU_INFO" | grep -Eq "HD Graphics ([4-9][0-9]{2}|[1-9][0-9]{3,})|Iris|Xe"; then
                    echo "Your Intel GPU should support Cosmic Desktop."
                else
                    echo "Older Intel GPU detected. Only HD Graphics 4000 and newer support Cosmic Desktop."
                fi
                ;;
            *AMD*)
                if echo "$GPU_INFO" | grep -Eq "RX (4[8-9][0-9]|[5-9][0-9]{2,})|VEGA|RDNA|RADEON PRO"; then
                    echo "Your AMD GPU should support Cosmic Desktop."
                else
                    echo "Older AMD GPU detected. Only RX 480 and newer support Cosmic Desktop."
                fi
                ;;
            *)
                echo "Unknown or unsupported GPU detected. Compatibility uncertain."
                ;;
        esac
    else
        echo "Cannot detect GPU. 'lspci' command not found."
    fi
}
```
All of this is redundant.

fam007e replied:
Shouldn't the user know whether they can install it or not...

nnyyxxxx replied:
@fam007e Please source where you found these specifications.

nnyyxxxx commented on src/commands/system-setup/arch/cosmic-setup.sh:
```diff
@@ -0,0 +1,133 @@
+                ;;
+        esac
+    fi
+}
```
```bash
# Check GPU compatibility for Wayland and X11
check_gpu_cosmic_compatibility() {
    if command -v lspci >/dev/null 2>&1; then
        GPU_INFO=$(lspci | grep -E "VGA|3D")

        case "$GPU_INFO" in
            *NVIDIA*)
                if echo "$GPU_INFO" | grep -Eq "GTX (9[0-9]{2}|[1-9][0-9]{3})|RTX|Titan|A[0-9]{2,3}"; then
                    echo "Your NVIDIA GPU should support both Wayland and X11."
                else
                    echo "Older NVIDIA GPU detected. Only 900 series and newer support Cosmic Desktop on both Wayland and X11."
                fi
                ;;
            *Intel*)
                if echo "$GPU_INFO" | grep -Eq "HD Graphics ([4-9][0-9]{2}|[1-9][0-9]{3,})|Iris|Xe"; then
                    echo "Your Intel GPU should support both Wayland and X11."
                else
                    echo "Older Intel GPU detected. Only HD Graphics 4000 and newer support Cosmic Desktop on both Wayland and X11."
                fi
                ;;
            *AMD*)
                if echo "$GPU_INFO" | grep -Eq "RX (4[8-9][0-9]|[5-9][0-9]{2,})|VEGA|RDNA|RADEON PRO"; then
                    echo "Your AMD GPU should support both Wayland and X11."
                else
                    echo "Older AMD GPU detected. Only RX 480 and newer support Cosmic Desktop on both Wayland and X11."
                fi
                ;;
            *)
                echo "Unknown or unsupported GPU detected. Compatibility uncertain."
                ;;
        esac
    fi
}
```
All of this is still redundant and should be removed.

fam007e replied:
[SRC](https://github.com/xerolinux/xero-plasma/blob/main/xero-cosmic.sh)

fam007e replied:
I also saw a few reddit post about some driver issues but can't really them so I am going remove it as per your guidance.

ChrisTitusTech commented:
Sorry for the inconvenience. We had a massive restructure of the codebase to improve future development. Because of this can you update your PR to the new structure. Thank you for your assistance and contribution.

nnyyxxxx commented:
@fam007e This needs to be rebased.

fam007e commented:
❯ git pull origin main
From github.com:fam007e/linutil
 * branch            main       -> FETCH_HEAD
Already up to date.
❯ git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean
❯ git rebase origin/main
Current branch main is up to date.
```
Despite a successful rebase and a clean status, GitHub still indicates that there are conflicts. I’m perplexed by this.

Here’s what I did:
- Cleared my entire repository and re-added the files to incorporate recent structural changes.
- Added my scripts and made minor edits to `hint.rs` to correct a few spelling mistakes.
- Updated the logo to an ASCII representation of "Linutil" and used ANSI color codes in `ratatui` to add color (blue) to both the "CTT" and the ASCII logo "Linutil" via changes in `state.rs`.
- Adjusted the `tabs/system-setup/tab_data.toml` file to accommodate the re-added scripts.
- Ran `cargo clean && cargo build && ./target/debug/linutil` to test my changes.

As a result, the image in the `README.md` was updated accordingly.

I would appreciate any guidance on resolving this issue.

jeevithakannan2 commented:
Can you please split this PR into separate feature PRs

fam007e commented:
> Can you please split this PR into separate feature PRs

I am going for the hard reset. I will make the fork again and make two batches of commits one is for the scripts and another for the `.rs` files.

jeevithakannan2 commented:
Each of the script should be a separate PR so that it will be easy to check
